var/spack/repos/builtin/packages/foam-extend/package.py

```diff
@@ -40,207 +130,234 @@ class FoamExtend(Package):
+        'WM_COMPILE_OPTION':  'SPACKOpt',  # Do not change
+        'WM_MPLIB':           'SPACK',     # Do not change
+    }
+
+    # The system description is frequently needed
+    foam_sys = {
+        'WM_ARCH':      None,
+        'WM_COMPILER':  None,
+        'WM_OPTIONS':   None,
+    }
+
+    build_script = './spack-Allwmake'  # <- Generated by patch() method.
+    # phases = ['configure', 'build', 'install']
 
-    depends_on('paraview', when='+paraview')
+    def setup_environment(self, spack_env, run_env):
+        run_env.set('FOAM_INST_DIR', self.prefix)
+        run_env.set('WM_PROJECT_DIR', self.projectdir)
+
+    @property
+    def _canonical(self):
+        """Canonical name for this package and version"""
+        return 'foam-extend-{0}'.format(self.version.up_to(2))
+
+    @property
+    def projectdir(self):
+        """Absolute location of project directory: WM_PROJECT_DIR/"""
+        return join_path(self.prefix, self._canonical)  # <- prefix/canonical
+
+    @property
+    def etc(self):
+        """Absolute location of the OpenFOAM etc/ directory"""
+        return join_path(self.projectdir, 'etc')
+
+    @property
+    def archbin(self):
+        """Relative location of architecture-specific executables"""
+        wm_options = self.set_openfoam()
+        return join_path('applications', 'bin', wm_options)
+
+    @property
+    def archlib(self):
+        """Relative location of architecture-specific libraries"""
+        wm_options = self.set_openfoam()
+        return join_path('lib', wm_options)
+
+    @property
+    def mplib_content(self):
+        """Define 'SPACK' mpi settings to have wmake
+        use spack information with minimum modifications to OpenFOAM
+        """
+        return [
+            'PFLAGS = -DOMPI_SKIP_MPICXX -DMPICH_IGNORE_CXX_SEEK',
+            'PINC   = -I{0}'.format(self.spec['mpi'].prefix.include),
+            'PLIBS  = -L{0} -lmpi'.format(pkglib(self.spec['mpi']))
+        ]
+
+    @property
+    def rpath_info(self):
+        """Define 'SPACKOpt' compiler optimization file to have wmake
+        use spack information with minimum modifications to OpenFOAM
+        """
+        build_libpath   = join_path(self.stage.source_path, self.archlib)
+        install_libpath = join_path(self.projectdir, self.archlib)
+
+        # 'DBUG': rpaths
+        return '{0}{1} {2}{3}'.format(
+            self.compiler.cxx_rpath_arg, install_libpath,
+            self.compiler.cxx_rpath_arg, build_libpath,
+        )
+
+    # For foam-extend: supports darwinIntel, does not yet support +int64
+    def set_openfoam(self):
+        """Populate foam_cfg, foam_sys according to
+        variants, architecture, compiler.
+        Returns WM_OPTIONS.
+        """
+        # Run once
+        wm_options = self.foam_sys['WM_OPTIONS']
+        if wm_options:
+            return wm_options
+
+        # Trivial check
+        if self.compiler.name not in self.supported_compilers:
```

Comment:
Is this necessary?  We generally leave things unconstrained in Spack to make it easy to build new versions of things.  If someone tries a strange compiler and it happens to work, is that bad?  People encounter this a lot when porting to new machines.